### PR TITLE
Add label filtering functionality to the dataset class

### PR DESCRIPTION
**Pull request by AlbertvanHouten**
This pull request introduces a new `filter_by_labels` method to the `Dataset` class, enabling users to filter datasets based on label values. The implementation is robust, supporting a variety of label field configurations (single label, multi-label, list fields, and hierarchical categories), and includes comprehensive validation and error handling. Extensive unit tests have been added to verify functionality, edge cases, and error conditions.

**New filtering functionality:**

- Added the `filter_by_labels` method to the `Dataset` class, allowing filtering of dataset items by label names or indices. The method supports auto-detection of the label field when the schema contains exactly one `LabelField`, and works with both `LabelCategories` and `HierarchicalLabelCategories`.

Required for https://github.com/open-edge-platform/training_extensions/issues/5564

### Checklist

- [ ] I have added tests to cover my changes or documented any manual tests.
- [ ] I have updated the [documentation](https://github.com/open-edge-platform/datumaro/tree/develop/docs) accordingly

### CODE REVIEW

File: src/datumaro/experimental/dataset.py [hunk 1]
```diff
@@ -599,6 +601,300 @@ def filter_by_subset(self, subset: Subset | Sequence[Subset]) -> Dataset[DType]:
...
+
+        attr_info = self.schema.attributes[label_field_name]
+        if not isinstance(attr_info.field, LabelField):
+            raise TypeError(f"Attribute '{label_field_name}' is a {type(attr_info.field).__name__}, not a LabelField.")
+
+        return label_field_name
+
+    @staticmethod
+    def _resolve_label_indices(
+        labels: Sequence[str | int],
```
Comment (JortBergfeld) [inline src/datumaro/experimental/dataset.py]: I suppose it should be `Sequence[int] | Sequence[str]`, to ensure we do not have to deal with the duplicate label problem described above by copilot

File: src/datumaro/experimental/dataset.py [hunk 2]
```diff
@@ -599,6 +601,300 @@ def filter_by_subset(self, subset: Subset | Sequence[Subset]) -> Dataset[DType]:
...
+    ) -> pl.DataFrame:
+        """Return a filtered DataFrame keeping rows that match any of *label_indices*.
+
+        This method both filters rows (keeping only those with at least one matching label)
+        AND removes non-matching labels from within each sample's label field.
+
+        The filtering strategy is chosen based on the ``is_list`` and
+        ``multi_label`` flags of *label_field_instance*.
+        """
+        col = pl.col(label_field_name)
```
Comment (JortBergfeld) [inline src/datumaro/experimental/dataset.py]: Prefer to call this `label_col` as it is easy to confuse `col` with the Polars method.

File: src/datumaro/experimental/dataset.py [hunk 3]
```diff
@@ -599,6 +601,300 @@ def filter_by_subset(self, subset: Subset | Sequence[Subset]) -> Dataset[DType]:
...
+        Args:
+            df: The DataFrame to modify
+            label_field_name: Name of the label column
+            label_field_instance: The LabelField instance
+            old_to_new_index_map: Mapping from old indices to new indices
+
+        Returns:
+            DataFrame with remapped label indices
+        """
+        col = pl.col(label_field_name)
```
Comment (JortBergfeld) [inline src/datumaro/experimental/dataset.py]: Similar here, I prefer `label_col`

File: src/datumaro/experimental/dataset.py [hunk 4]
```diff
@@ -599,6 +601,300 @@ def filter_by_subset(self, subset: Subset | Sequence[Subset]) -> Dataset[DType]:
...
+        Example:
+            ```python
+            # Auto-detect (schema has exactly one LabelField)
+            filtered = dataset.filter_by_labels(["cat", "dog"])
+
+            # Using label indices
+            filtered = dataset.filter_by_labels([0, 1])
+
+            # Mixing label names and indices
+            filtered = dataset.filter_by_labels(["cat", 1])
```
Comment (JortBergfeld) [inline src/datumaro/experimental/dataset.py]: I'm not sure why we would support this and certainly not why we would promote its usage.

**Review by JortBergfeld — COMMENTED** [review]
Looks good, but in general I feel that we are adding too much code to the dataset class for this one feature. Perhaps we can split at least the private helper methods into a different module/class, or everything and just keep a helper method that calls the other class. After all, it returns a new dataframe, it doesn't update it, so it could be an independent function.

BTW, my comments are minor, but not approving as I am waiting for your response to copilot's comments

File: src/datumaro/experimental/dataset.py (hunk 4 above)
Comment (AlbertvanHouten) [reply src/datumaro/experimental/dataset.py]: I've removed the example.

**Review by JortBergfeld — APPROVED** [review]
